### PR TITLE
net: l2: ieee802154: improved/fixed endianness and short address handling

### PR DESCRIPTION
**Pull request by ghost**
This PR implements part one of the change announced in https://github.com/zephyrproject-rtos/zephyr/issues/50336#issuecomment-1250250154:

> Improve short address support in the IEEE 802.15.4 stack as a precondition for TSCH link tables (currently short addresses are not fully supported)

It turned out that a neighbor table was not (yet) required for this preparatory change. The PR also fixes endianness in the IEEE 802.15.4 L2 stack as the work on this change revealed several inconsistencies and even bugs in byte order conversion.

### CODE REVIEW

File: include/zephyr/net/ieee802154_mgmt.h [hunk 1]
```diff
@@ -42,14 +42,14 @@ enum net_request_ieee802154_cmd {
...
-	NET_REQUEST_IEEE802154_CMD_GET_PAN_ID,
-	NET_REQUEST_IEEE802154_CMD_SET_EXT_ADDR,
-	NET_REQUEST_IEEE802154_CMD_GET_EXT_ADDR,
-	NET_REQUEST_IEEE802154_CMD_SET_SHORT_ADDR,
-	NET_REQUEST_IEEE802154_CMD_GET_SHORT_ADDR,
+	NET_REQUEST_IEEE802154_CMD_SET_CHANNEL, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_GET_CHANNEL, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_SET_PAN_ID, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_GET_PAN_ID, /* in CPU byte order */
+	NET_REQUEST_IEEE802154_CMD_SET_EXT_ADDR, /* in big endian byte order */
```
Comment (rlubos) [inline include/zephyr/net/ieee802154_mgmt.h:49]: The commit message in `net: l2: ieee802154: document and fix endianness` mentions "external address" twice - shouldn't it rather be "extended address"?

File: include/zephyr/net/ieee802154_radio.h [hunk 2]
```diff
@@ -271,24 +271,29 @@ struct ieee802154_config {
 		} rx_slot;
 
 		/** ``IEEE802154_CONFIG_CSL_PERIOD`` */
-		uint32_t csl_period;
+		uint32_t csl_period; /* in CPU byte order */
 
 		/** ``IEEE802154_CONFIG_CSL_RX_TIME`` */
-		uint32_t csl_rx_time;
+		uint32_t
+			csl_rx_time; /* in microseconds, based on ieee802154_radio_api.get_time() */
```
Comment (rlubos) [inline include/zephyr/net/ieee802154_radio.h:278]: Why such an unexpected line break? IMO better to break the comment if it does not fit.

File: subsys/net/ip/net_context.c [hunk 3]
```diff
@@ -758,17 +758,17 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
...
+			if (ll_addr->sll_halen != 0) {
 				return -EINVAL;
 			}
-			net_sll_ptr(&context->local)->sll_addr =
-				ll_addr->sll_addr;
-			net_sll_ptr(&context->local)->sll_halen =
-				ll_addr->sll_halen;
-		} else {
+
+#ifdef CONFIG_NET_L2_IEEE802154
```
Comment (rlubos) [inline subsys/net/ip/net_context.c:766]: It's already within `IS_ENABLED(CONFIG_NET_L2_IEEE802154)` block, so this #ifdef seems redundant?

**Review by rlubos — COMMENTED** [review]
Had a brief look, not much to complain, just a few nits. Looks pretty good!

File: include/zephyr/net/ieee802154_mgmt.h (hunk 1 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154_mgmt.h:49]: Oups. Of course. This was just a test whether you actually read my commit messages. ;-)

File: include/zephyr/net/ieee802154_radio.h (hunk 2 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154_radio.h:278]: Fixed.

File: subsys/net/ip/net_context.c (hunk 3 above)
Comment (fgrandel) [reply subsys/net/ip/net_context.c]: Fortunately I could solve this differently now so no need to mention a preprocessor constant at all.

The problem was, that the code wouldn't compile without `#ifdef` as `IS_ENABLED()` will let the compiler see the code inside the block which only works if all code referenced inside this block is also visible to the compiler. This wasn't the case for some struct field in this block which was defined conditionally. Using `IS_ENABLED()` for struct fields wastes memory, though, that the compiler cannot optimize away AFAICS. So while I quite like the `IS_ENABLED()` pattern now that I fully understand its purpose it doesn't seem to work in such cases.

**Comment by fgrandel** [comment]
@rlubos : Thanks for the review! I also fixed the remaining bugs today, that made the tests fail and I added extra test coverage for short address mode. So should be ready for final review.

**Review by rlubos — DISMISSED** [review]
LGTM

**Review by jukkar — DISMISSED** [review]
LGTM

**Review by jciupis — DISMISSED** [review]

File: include/zephyr/net/ieee802154.h [hunk 4]
```diff
@@ -42,6 +42,8 @@ extern "C" {
 
 /* See IEEE 802.15.4-2006, section 7.2.1.4 */
 #define IEEE802154_BROADCAST_ADDRESS 0xFFFF
+#define IEEE802154_NO_SHORT_ADDRESS_ASSIGNED 0xFFFE
```
Comment (tbursztyka) [inline include/zephyr/net/ieee802154.h:45]: Split the commit. It's too many things in one, hard to follow.
- one about the short address it 802.15.4
- one for context
- one about the association/disassociation stuff

File: subsys/net/ip/net_context.c [hunk 5]
```diff
@@ -763,15 +763,25 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
 				return -EINVAL;
 			}
 
+			struct ieee802154_context *ieee802154_context =
```
Comment (tbursztyka) [inline subsys/net/ip/net_context.c:766]: No, no l2 stuff into net_context.c
You'll have to find another way.

File: subsys/net/l2/ieee802154/ieee802154.c [hunk 6]
```diff
@@ -380,9 +387,12 @@ static int ieee802154_enable(struct net_if *iface, bool state)
 
 	NET_DBG("iface %p %s", iface, state ? "up" : "down");
 
+	k_sem_take(&ctx->ctx_lock, K_FOREVER);
```
Comment (tbursztyka) [inline subsys/net/l2/ieee802154/ieee802154.c:390]: Please add empty lines: 
- before the if (unless its statement is about what has been computed before), linux code style
- after the '}' (unless it's another '}') etc..

Apply that everywhere

File: subsys/net/l2/ieee802154/ieee802154_frame.c [hunk 7]
```diff
@@ -916,19 +936,24 @@ bool ieee802154_create_ack_frame(struct net_if *iface, struct net_pkt *pkt, uint
 bool ieee802154_decipher_data_frame(struct net_if *iface, struct net_pkt *pkt,
 				    struct ieee802154_mpdu *mpdu)
 {
+	bool ret = false;
```
Comment (tbursztyka) [inline subsys/net/l2/ieee802154/ieee802154_frame.c:945]: try to keep a reversed christmas tree variable ordering when applicable.

File: subsys/net/l2/ieee802154/ieee802154_mgmt.c [hunk 8]
```diff
@@ -150,28 +157,30 @@ static int ieee802154_scan(uint32_t mgmt_request, struct net_if *iface,
 			if (ret) {
 				NET_DBG("Could not send Beacon Request (%d)",
 					ret);
-				net_pkt_unref(pkt);
```
Comment (tbursztyka) [inline subsys/net/l2/ieee802154/ieee802154_mgmt.c:152]: No, the packet got referenced above: line 153.

File: subsys/net/l2/ieee802154/ieee802154_mgmt.c [hunk 9]
```diff
@@ -258,26 +272,34 @@ enum net_verdict ieee802154_handle_mac_command(struct net_if *iface,
 			return NET_DROP;
 		}
 
+		k_sem_take(&ctx->ctx_lock, K_FOREVER);
+
+		enum net_verdict ret = NET_DROP;
```
Comment (tbursztyka) [inline subsys/net/l2/ieee802154/ieee802154_mgmt.c:277]: kind of ugly to see a variable declaration it a middle of a code block

File: include/zephyr/net/ieee802154.h [hunk 10]
```diff
@@ -85,11 +82,15 @@ struct ieee802154_context {
...
-	uint8_t _unused		: 5;
+	struct k_sem ctx_lock; /* guards all mutable context attributes unless
+				* otherwise mentioned on attribute level
+				*/
+
+	uint8_t ack_seq;	   /* guarded by ack_lock */
+	uint8_t ack_received : 1;  /* guarded by ack_lock */
+	uint8_t ack_requested : 1; /* guarded by ack_lock */
+	uint8_t _unused : 6;
+	struct k_sem ack_lock;
```
Comment (tbursztyka) [inline include/zephyr/net/ieee802154.h:93]: looks like reordering the attributes this way, you mess up the structure alignment. (thus probably lose some space)

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Overall a good PR, but some commits are too big and include too many changes at once: please split them. Side note: try to keep consistent with the code style.

**Comment by fgrandel** [comment]
> Overall a good PR, but some commits are too big and include too many changes at once: please split them. Side note: try to keep consistent with the code style.

Thank you @tbursztyka for this thorough review. Will look into it this weekend to address your concerns.

File: subsys/net/l2/ieee802154/ieee802154_mgmt.c (hunk 8 above)
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154_mgmt.c:152]: This code block now calls the "out" label where the packet is being unrefed. Therefore this should be ok.

File: subsys/net/ip/net_context.c (hunk 5 above)
Comment (fgrandel) [reply subsys/net/ip/net_context.c:766]: This led me to better investigate the locking strategy of the ll address in net_if. I'm now relying on net_if's own lock which is much better. No need to do any L2-specific locking to update the ll-address any more. Also reminded me to cover the corresponding ieee802154_mgmt API in the test which again revealed a bug. Thanks for this useful remark.

File: subsys/net/l2/ieee802154/ieee802154_mgmt.c (hunk 8 above)
Comment (tbursztyka) [reply subsys/net/l2/ieee802154/ieee802154_mgmt.c:152]: No, it got referenced once more just 7 lines before, we need to get rid of this reference.
When the pkt gets created, its reference count has been initialized to 1. Here in this code block we put it to 2. 
If we specifically do not get back to a ref count of 1, at the out label it will get back to 1 and... the pkt will not be freed. (it needs to get to 0 for this to happen). So this pkt will be accounted as memory loss.

File: drivers/ieee802154/ieee802154_nrf5.c [hunk 11]
```diff
@@ -940,9 +940,10 @@ static int nrf5_configure(const struct device *dev,
 		sys_put_le16(config->ack_ie.short_addr, short_addr_le);
 		/**
 		 * The extended address field passed to this function starts
-		 * with the leftmost octet and ends with the rightmost octet.
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply drivers/ieee802154/ieee802154_nrf5.c:943]: Fixed.

File: subsys/net/ip/connection.c [hunk 12]
```diff
@@ -760,8 +760,7 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
 		} else if (IS_ENABLED(CONFIG_NET_SOCKETS_CAN) && conn_family == AF_CAN) {
 			best_match = conn;
 		}
-			/* loop end */
-	}
+	} /* loop end */
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply subsys/net/ip/connection.c:763]: Really? For a one-liner? Seems like a lot of maintenance overhead for the maintainers (and myself). As an external contributor I'd stop fixing such things if they always had to be done in separate PRs. (While a separate commit is totally fine of course.) If you insist I'll do so but it would be great if you could extend a little on your rationale behind it so it becomes easier to accept this.

File: include/zephyr/net/net_ip.h [hunk 13]
```diff
@@ -201,23 +201,23 @@ struct sockaddr_in_ptr {
...
 /** Socket address struct for packet socket. */
 struct sockaddr_ll {
-	sa_family_t sll_family;   /* Always AF_PACKET        */
-	uint16_t    sll_protocol; /* Physical-layer protocol */
-	int         sll_ifindex;  /* Interface number        */
-	uint16_t    sll_hatype;   /* ARP hardware type       */
-	uint8_t     sll_pkttype;  /* Packet type             */
-	uint8_t     sll_halen;    /* Length of address       */
-	uint8_t     sll_addr[8];  /* Physical-layer address  */
+	sa_family_t sll_family;   /* Always AF_PACKET                   */
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/net/net_ip.h:204]: This is closely related to other stuff in this PR. So while I definitely agree that this should go into a separate commit so that maintainers can pick commits relevant to their code ownership I do not yet understand the point of creating separate PRs (but I'm of course open to change my mind with a good argument). As in the other comment above: If you insist I'll of course follow your advice but would be great if you could give the "why?" together with the "what?".

File: include/zephyr/net/net_linkaddr.h [hunk 14]
```diff
@@ -66,7 +66,7 @@ enum net_link_type {
  */
 struct net_linkaddr {
 	/** The array of byte representing the address */
-	uint8_t *addr;
+	uint8_t *addr; /* in big endian */
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/net/net_linkaddr.h:69]: Duplicate, see discussion above. Went into a different commit.

File: include/zephyr/net/ieee802154_radio.h [hunk 15]
```diff
@@ -313,7 +319,7 @@ struct ieee802154_radio_api {
 	int (*cca)(const struct device *dev);
 
 	/** Set current channel */
-	int (*set_channel)(const struct device *dev, uint16_t channel);
+	int (*set_channel)(const struct device *dev, uint16_t channel /* in CPU byte order */);
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/net/ieee802154_radio.h:322]: Fixed.

File: subsys/net/l2/ieee802154/ieee802154_frame.h [hunk 16]
```diff
@@ -114,8 +114,8 @@ struct ieee802154_fcf_seq {
 struct ieee802154_address {
 	union {
 		uint8_t simple_addr;
-		uint16_t short_addr;
-		uint8_t ext_addr[0];
+		uint16_t short_addr; /* in little endian */
```
[the remark this answers is not in the document]
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154_frame.h:118]: Fixed.

File: include/zephyr/net/ieee802154.h (hunk 4 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154.h:45]: Fixed.

File: subsys/net/l2/ieee802154/ieee802154.c (hunk 6 above)
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154.c:390]: Fixed here and in all other places throughout the whole change set.

File: subsys/net/l2/ieee802154/ieee802154_frame.c (hunk 7 above)
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154_frame.c:945]: Fixed, but where can I find this rule (and others you claimed) documented? I looked in the [coding guidelines](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) and my code is passing all automatic style checks. If such rules are expected by the project why are they not enforced by the style checker and/or documented? Or did I simply overlook them?

File: subsys/net/l2/ieee802154/ieee802154_mgmt.c (hunk 9 above)
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154_mgmt.c:277]: Fixed.

File: include/zephyr/net/ieee802154.h (hunk 10 above)
Comment (fgrandel) [reply include/zephyr/net/ieee802154.h:93]: Good point. Not the re-ordering was problematic but the introduction of the additional semaphore in the wrong place. Fixed this. There was also additional pre-existing padding in the struct (16 bytes overall) that could be removed without impacting readability. This is done in a separate commit now.

**Comment by fgrandel** [comment]
@tbursztyka I addressed all your concerns except for creating a separate PR for the IP subsystem - see my comments. It would be great if you could answer my questions in the unresolved comments and/or approve the PR.

File: subsys/net/l2/ieee802154/ieee802154_mgmt.c (hunk 8 above)
Comment (fgrandel) [reply subsys/net/l2/ieee802154/ieee802154_mgmt.c:152]: As the comment was hidden I overlooked your response. Will fix this now.

File: include/zephyr/net/net_ip.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/zephyr/net/net_ip.h:204]: A split commit will do

**Review by tbursztyka — APPROVED** [review]

**Comment by fgrandel** [comment]
@tbursztyka Thanks for your valuable feedback and approval.

@rlubos @jukkar Another approval seems to be required to unblock this PR. You already posted LGTM, seems like its just a missing button press?

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
